File: poolutilityjava/src/main/java/jgappsandgames/me/poolutilityjava/Pool.java
package jgappsandgames.me.poolutilityjava;

import java.util.ArrayList;

public class Pool<T extends PoolObjectInterface> {
    private ArrayList<T> pool;

    private int minSize;
    private int maxSize;
    private int increment;
    private int decrement;

    private PoolObjectCreator<T> generator;

    // Initializer ---------------------------------------------------------------------------------
    public Pool(Integer minSize, Integer maxSize, Integer increment, Integer decrement, PoolObjectCreator<T> generator) {
        if (minSize == null) this.minSize = 10;
        else this.minSize = minSize;

        if (maxSize == null) this.maxSize = Integer.MAX_VALUE;
        else this.maxSize = maxSize;

        if (increment == null) this.increment = this.minSize;
        else this.increment = increment;

        if (decrement == null) this.decrement = this.increment / 2;
        else this.decrement = decrement;

        this.generator = generator;

        pool = new ArrayList<>(this.minSize);
        for (int i = 0; i < this.minSize; i++) {
            pool.add(generator.generatePoolObject());
        }
    }

    // Management ----------------------------------------------------------------------------------
    public void changeMin(int min) {
        minSize = min;
    }

    public void changeMax(int max) {
        maxSize = max;
    }

    public void changeIncrement(int increment) {
        this.increment = increment;
    }

    public void changeDecrement(int decrement) {
        this.decrement = decrement;
    }

    public void lowMemory() {
        if (pool.size() > minSize) {
            for (int i = 0; i < decrement; i++) {
                pool.remove(0);
                if (pool.size() <= minSize) break;
            }
        }

        for (int i = 0; i < pool.size(); i++) {
            pool.get(i).deconstruct();
        }
    }

    // Pool Handlers -------------------------------------------------------------------------------
    public void enlargePool() {
        for (int i = 0; i < increment; i++) pool.add(generator.generatePoolObject());
    }

    public void enlargePool(int amount) {
        for (int i = 0; i < amount; i++) pool.add(generator.generatePoolObject());
    }

    public T getPoolObject() {
        if (pool.size() == 0) enlargePool();

        return pool.remove(0);
    }

    public void returnPoolObject(T poolObject) {
        returnPoolObject(poolObject, true);
    }

    public void returnPoolObject(T poolObject, boolean deconstruct) {
        if (deconstruct) poolObject.deconstruct();
        pool.add(poolObject);
    }
}

File: poolutilitykotlin/src/main/java/jgappsandgames/me/poolutilitykotlin/Pool.kt
package jgappsandgames.me.poolutilitykotlin

class Pool<T: PoolObjectInterface>(private var minSize: Int = 10, private var maxSize: Int = Int.MAX_VALUE,
                                   private var increment: Int = minSize, private var decrement: Int = (increment / 2),
                                   private val generator: PoolObjectCreator<T>) {
    private val pool: ArrayList<T> = ArrayList(minSize)

    // Initializer ---------------------------------------------------------------------------------
    init {
        for (i in 0 until minSize) pool.add(generator.generatePoolObject())
    }

    // Management Methods --------------------------------------------------------------------------
    fun changeMin(min: Int) {
        minSize = min
    }

    fun changeMax(max: Int) {
        maxSize = max
    }

    fun changeIncrement(increment: Int) {
        this.increment = increment
    }

    fun changeDecrement(decrement: Int) {
        this.decrement = decrement
    }

    fun lowMemory() {
        if (pool.size > minSize) {
            for (i in 0 until decrement) {
                pool.removeAt(0)
                if (pool.size <= minSize) break
            }
        }

        for (i in 0 until pool.size) {
            pool[i].deconstruct()
        }
    }

    // Pool Handlers -------------------------------------------------------------------------------
    fun enlargePool(amount: Int = increment) {
        for (i in 0 until amount) pool.add(generator.generatePoolObject())
    }

    fun getPoolObject(): T {
        if (pool.size == 0) enlargePool()

        return pool.removeAt(0)
    }

    fun returnPoolObject(poolObject: T, deconstruct: Boolean = true) {
        if (deconstruct) poolObject.deconstruct()
        pool.add(poolObject)
    }
}